File: backends/java/zappy-domain/src/test/java/com/zappymart/domain/promotions/PromotionTest.java
package com.zappymart.domain.promotions;

import com.zappymart.domain.shared.Money;
import com.zappymart.domain.shared.Result;
import com.zappymart.domain.shared.UserError;
import com.zappymart.domain.shared.UserErrorCode;
import org.junit.jupiter.api.Test;

import java.time.Instant;

import static org.assertj.core.api.Assertions.assertThat;

class PromotionTest {

    private static final Instant TODAY = Instant.parse("2026-09-09T12:00:00Z");
    private static final Instant OPENS = Instant.parse("2026-01-01T00:00:00Z");
    private static final Instant CLOSES = Instant.parse("2027-12-31T23:59:59Z");

    @Test
    void aPercentageCodeTakesItsShareOfTheSubtotal() {
        Promotion welcome = percentage("WELCOME10", 10, null, OPENS, CLOSES, null, 0);

        AppliedPromotion applied = welcome.applyTo(Money.euro(5599), TODAY).valueOrThrow();

        assertThat(applied.code()).isEqualTo("WELCOME10");
        assertThat(applied.kind()).isEqualTo(PromotionKind.PERCENTAGE);
        assertThat(applied.discount()).isEqualTo(Money.euro(560));
    }

    @Test
    void aCodeOutsideItsWindowIsExpired() {
        Promotion summer = percentage("SUMMER2025", 10, null,
                Instant.parse("2025-06-01T00:00:00Z"), Instant.parse("2025-08-31T23:59:59Z"), null, 0);

        assertThat(reasonOf(summer.applyTo(Money.euro(5599), TODAY))).isEqualTo(UserErrorCode.CODE_EXPIRED);
    }

    @Test
    void aCodeAtItsLimitIsExhausted() {
        Promotion once = percentage("ONCE", 10, null, OPENS, CLOSES, 1, 1);

        assertThat(once.isExhausted()).isTrue();
        assertThat(reasonOf(once.applyTo(Money.euro(5599), TODAY))).isEqualTo(UserErrorCode.CODE_EXHAUSTED);
    }

    @Test
    void aCodeBelowItsMinimumSubtotalIsRefused() {
        Promotion fiveOff = new Promotion(new PromotionCode("FIVEOFF"),
                new PromotionRule.FixedAmountOff(Money.euro(500)), Money.euro(2500), OPENS, CLOSES, null, 0);

        assertThat(reasonOf(fiveOff.applyTo(Money.euro(1970), TODAY)))
                .isEqualTo(UserErrorCode.CODE_MINIMUM_NOT_MET);
        assertThat(fiveOff.applyTo(Money.euro(2500), TODAY).valueOrThrow().discount())
                .isEqualTo(Money.euro(500));
    }

    @Test
    void aFreeShippingCodeDiscountsNothingAndCarriesTheShipping() {
        Promotion freeShipping = new Promotion(new PromotionCode("FREESHIP"),
                new PromotionRule.FreeShipping(), null, OPENS, CLOSES, null, 0);

        AppliedPromotion applied = freeShipping.applyTo(Money.euro(1970), TODAY).valueOrThrow();

        assertThat(applied.discount()).isEqualTo(Money.zero());
        assertThat(applied.carriesShipping()).isTrue();
    }

    @Test
    void countsOneMoreUse() {
        Promotion once = percentage("ONCE", 10, null, OPENS, CLOSES, 2, 0);

        assertThat(once.afterOneMoreUse().timesUsed()).isEqualTo(1);
        assertThat(once.minimum()).isEmpty();
    }

    private static Promotion percentage(String code, int percentage, Money minimumSubtotal,
                                        Instant from, Instant until, Integer limit, int used) {
        return new Promotion(new PromotionCode(code), new PromotionRule.PercentageOff(percentage),
                minimumSubtotal, from, until, limit, used);
    }

    private static UserErrorCode reasonOf(Result<AppliedPromotion> result) {
        return result.errors().stream().map(UserError::code).findFirst().orElseThrow();
    }
}

File: backends/kotlin/zappy-domain/src/test/kotlin/nl/zappymart/domain/promotions/PromotionTest.kt
package nl.zappymart.domain.promotions

import java.time.Instant
import nl.zappymart.domain.shared.Money
import nl.zappymart.domain.shared.PromotionCode
import nl.zappymart.domain.shared.Result
import nl.zappymart.domain.shared.UserErrorCode
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class PromotionTest {

    private val today: Instant = Instant.parse("2026-09-09T10:00:00Z")

    private fun promotion(
        code: String,
        rule: PromotionRule,
        minimumSubtotal: Money? = null,
        validFrom: String = "2026-01-01T00:00:00Z",
        validUntil: String = "2027-12-31T23:59:59Z",
        usageLimit: Int? = null,
        timesUsed: Int = 0,
    ) = Promotion(
        PromotionCode.of(code),
        rule,
        minimumSubtotal,
        Instant.parse(validFrom),
        Instant.parse(validUntil),
        usageLimit,
        timesUsed,
    )

    @Test
    fun `a percentage code takes its percentage of the subtotal rounded half up`() {
        val applied = promotion("WELCOME10", PromotionRule.Percentage(10))
            .applyTo(Money.euro(5599), today) as Result.Success
        assertThat(applied.value.kind).isEqualTo(PromotionKind.PERCENTAGE)
        assertThat(applied.value.discount).isEqualTo(Money.euro(560))
    }

    @Test
    fun `a fixed amount code never discounts more than the subtotal`() {
        val applied = promotion("FIVEOFF", PromotionRule.FixedAmount(Money.euro(500)))
            .applyTo(Money.euro(300), today) as Result.Success
        assertThat(applied.value.discount).isEqualTo(Money.euro(300))
    }

    @Test
    fun `a free shipping code discounts nothing`() {
        val applied = promotion("FREESHIP", PromotionRule.FreeShipping)
            .applyTo(Money.euro(1970), today) as Result.Success
        assertThat(applied.value.kind).isEqualTo(PromotionKind.FREE_SHIPPING)
        assertThat(applied.value.discount).isEqualTo(Money.NOTHING)
    }

    @Test
    fun `a code outside its window is expired`() {
        val refused = promotion(
            "SUMMER2025",
            PromotionRule.Percentage(10),
            validFrom = "2025-06-01T00:00:00Z",
            validUntil = "2025-08-31T23:59:59Z",
        ).applyTo(Money.euro(1970), today) as Result.Refused
        assertThat(refused.errors.map { error -> error.code }).containsExactly(UserErrorCode.CODE_EXPIRED)
    }

    @Test
    fun `a code at its usage limit is exhausted`() {
        val refused = promotion("ONCE", PromotionRule.Percentage(10), usageLimit = 1, timesUsed = 1)
            .applyTo(Money.euro(1970), today) as Result.Refused
        assertThat(refused.errors.map { error -> error.code }).containsExactly(UserErrorCode.CODE_EXHAUSTED)
    }

    @Test
    fun `a code below its minimum subtotal is refused`() {
        val refused = promotion(
            "FIVEOFF",
            PromotionRule.FixedAmount(Money.euro(500)),
            minimumSubtotal = Money.euro(2500),
        ).applyTo(Money.euro(1970), today) as Result.Refused
        assertThat(refused.errors.map { error -> error.code }).containsExactly(UserErrorCode.CODE_MINIMUM_NOT_MET)
    }

    @Test
    fun `counting a use raises the recorded number`() {
        assertThat(promotion("ONCE", PromotionRule.Percentage(10)).usedOnce().timesUsed).isEqualTo(1)
    }
}
